x/upgrade/keeper/keeper.go

```diff
@@ -100,19 +107,45 @@ func (k Keeper) SetUpgradedClient(ctx sdk.Context, cs ibcexported.ClientState) e
 	}
 
 	store := ctx.KVStore(k.storeKey)
-	store.Set(types.UpgradedClientKey(), bz)
+	store.Set(types.UpgradedClientKey(upgradeHeight), bz)
 	return nil
 }
 
 // GetUpgradedClient gets the expected upgraded client for the next version of this chain
-func (k Keeper) GetUpgradedClient(ctx sdk.Context) (ibcexported.ClientState, error) {
+// along with the planned upgrade height
+func (k Keeper) GetUpgradedClient(ctx sdk.Context) (ibcexported.ClientState, int64, error) {
 	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.UpgradedClientKey())
-	if len(bz) == 0 {
-		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "no upgraded client in store")
+	iterator := sdk.KVStorePrefixIterator(store, []byte(types.KeyUpgradedClient))
```

Comment:
I don't understand why are you using a prefix iterator instead of the key with the height?

Reply:
I want to be able to get the upgraded client in the store without knowing the upgrade height beforehand

Reply:
What is the use case for this? When do you not know the key?

Reply:
If the iterator is added to do a sanity check that the number of upgrades client's set is <= 1, then I'd suggest leaving `GetUpgradeClient` as it was with the height param provided and adding a separate function `GetAllUpgradeClients`. I think the usage would make more sense be more readable and more predicatable

Then you can check len(GetAllUpgradeClients) and take action accordingly

Reply:
I prefer to keep this as is, the reason being is that there should only ever be 1 upgraded client in the store at any given time. Any violation of that would imply a faulty implementation that would completely break the upgrade feature's correctness.

This function is only used for testing purposes, where it is useful for me to just check the upgraded client in the store without knowing its height beforehand

Reply:
Reworked this file to ensure there is never more than 1 upgrade client in the store.

The panic in GetUpgradedClient is a useful sanity check

Reply:
why don't we return error instead?

Reply:
> why don't we return error instead?

I think multiple upgrade clients is basically catastrophic for light client security. We don't want an error simply showing up in the logs, we want the network to halt and fix the issue

Reply:
can you update godoc to explain this reasoning and why it iterates rather than taking in the key?

Reply:
yes, please document why we panic 👍